### ISSUE TITLE
Improving diagonal handling

### ISSUE THREAD

**Issue by kbarnhart** [landlab/landlab#874]
As part of Landlab 2.0 planning, we think we should carefully consider the treatment of diagonals. 

This Issue starts to describe some of the points we've discussed on a call. Please edit and comment as we iterate on our solution. @mcflugen @SiccarPoint @gregtucker @nicgaspar @erkanistan @saisiddu @nathanlyons 

1) We should flesh out the adjacency structures associated with D8s and diagonals. Information about adjacency associated with D8s and diagonals should be private (leading underscore).

2) We need to be careful about the ``flow__link_to_receiver_node`` field. 

3) We think that there should be fields for links and diagonals.

4) We are not certain whether we think there should be fields storage on D8s. 
    Pros: Speed, continuity between values with the same name (e.g. discharge). 
    Cons: Challenge about duplication of information.

@SiccarPoint and @nicgaspar are working on digging into this for the next meeting.

**Comment by SiccarPoint** [comment]
OK, here's a possible approach that I think we're circling around. What's here I think is the Katy/Eric model as I think I've followed from our meetings, largely via KB. I like this!

- Diagonals, sensu stricto, are a good thing to have public facing. They, their geometric properties, and their connectivity should all be accessible and visible.
- It should probably be possible to store at a diagonal field on this basis (if you really want to).
- But, diagonals are inherently degenerate, in that they violate our graph theory framework, so...
- There needs to be a prominent thing in the docs somewhere that flags the issue to a user, explains our rationale for all this, and offers some advice on working with diagonals.
- It shouldn't be possible to store information on d8s. This kind of processing will need to be accommodated either through the user working explicitly with links-then-diagonals, and storing that info separately, or by creating local property storage through components that need it. This forbidding of storage is necessary to forbid duplication and divergence of data that could be stored e.g. both at links and d8s.

If we go this route, I might advocate we go a bit further, and would be keen to get thoughts on:

- Since links are legitimate graph structure but diagonals are degenerate, it is conceptually *really* weird to permit the grid to provide information that mixes the two. i.e., I'm proposing **the grid doesn't provide any d8 information at all**.

But, of course, having geometries specified this way might sometimes be really useful, if ideologically impure (burn the witch! burn the witch! etc etc) So, I would also propose:

- We provide a utility of some kind that allows these geometries to be generated off a grid without actually accessing grid properties per se. I can imagine this kind of syntax:

```
>>> from landlab import diagonals as dg
>>> mg = RasterModelGrid((3, 3), (1., 1.))
>>> dg.d8s_at_node(mg, ordering='links_then_diagonals')[mg.core_nodes]
array([[ 6,  8,  5,  3, 18, 17, 12, 15]])
>>> dg.d8s_at_node(mg, ordering='rotational')[mg.core_nodes]
array([[ 6, 18,  8, 17,  5, 12,  3, 15]])
```

This could of course be cleaner (i.e., ensure internal ordering consistency; permit storage of geometries that don't need to be recalculated) using an object of some kind:
```
>>> from landlab import diagonals as dg_obj
>>> mg = RasterModelGrid((3, 3), (1., 1.))
>>> myd8s = dg_obj(mg, ordering='links_then_diagonals')
>>> myd8s.d8s_at_node[mg.core_nodes]
array([[ 6,  8,  5,  3, 18, 17, 12, 15]])
```

The resulting d8_obj could then (perhaps) be a place where it would be allowed to store d8 data - but we'll have to think about that. Use of the object would be akin to signing away the insurance on a hire car... don't blame us if you do a bunch of damage to your code, you wanted the lambo and we gave it to you!

**Comment by nicgaspar** [comment]
After looking at `diagonals.py` and `flow_director_d8.py` here are some thoughts.

Currently if one instantiates a raster grid, some diagonal information is created and is available to the user. For example 
```
>>> rg = RasterModelGrid((3,4),1.)
>>> rg.number_of_diagonals
12
>>> rg. number_of_d8
29
```

There is also `nodes_at_d8`, `nodes_at_diagonals`, `d8_dirs_at_node` and so on. In other words, one gets the feeling that most of the information for links is also available for diagonals and d8s. 

But then that isn't always the case. For example, because there is `number_of_links_at_node`, I was expecting `number_of_diagonals_at_node` but that doesn't exist.

Also strange is how some of the information is given. For example, `d8_adjacent_nodes_at_node`. You might think that would be a matrix with the number of rows equal to the number of nodes, but you would be wrong! It has twice the number of rows as there are nodes.

Basically, it's a bit confusing that d8s and diagonals seems somewhat available and somewhat not. If I was a user and found the current situation, I would wonder why I can't do everything with diagonals and d8s that I can do with links. Also, if we are going to make some things visible to the user, why not make methods like `_calculate_gradients_at_d8_active_links` available too?

**Comment by gregtucker** [comment]
I agree with the @SiccarPoint outline above: banish D8s from the grid (but consider an "external" utility to handle some of their aspects).
